crates/ndc-query-plan/src/query_plan/requests.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::BTreeMap;
+
+use derivative::Derivative;
+use indexmap::IndexMap;
+use ndc_models::{self as ndc, RelationshipType};
+use nonempty::NonEmpty;
+
+use crate::{vec_set::VecSet, Type};
+
+use super::{Aggregate, ConnectorTypes, Expression, Field, Grouping, OrderBy};
+
+#[derive(Derivative)]
+#[derivative(
+    Clone(bound = ""),
+    Debug(bound = ""),
+    PartialEq(bound = "T::ScalarType: PartialEq")
+)]
+pub struct QueryPlan<T: ConnectorTypes> {
+    pub collection: ndc::CollectionName,
+    pub query: Query<T>,
+    pub arguments: BTreeMap<ndc::ArgumentName, Argument<T>>,
+    pub variables: Option<Vec<VariableSet>>,
+
+    /// Types for values from the `variables` map as inferred by usages in the query request. It is
+    /// possible for the same variable to be used in multiple contexts with different types. This
+    /// map provides sets of all observed types.
+    ///
+    /// The observed type may be `None` if the type of a variable use could not be inferred.
+    pub variable_types: VariableTypes<T::ScalarType>,
+
+    // TODO: type for unrelated collection
```

Comment:
Do we need a ticket here to track this?

Reply:
Maybe - I don't remember where this would come up